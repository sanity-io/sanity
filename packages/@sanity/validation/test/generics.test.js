const {Rule} = require('../src')

describe('generics', () => {
  test('should be able to construct an empty rule', () => {
    expect(() => new Rule()).not.toThrow(Error)
    expect(new Rule()).toBeInstanceOf(Rule)
  })

  test('should be able to construct a new typed rule', () => {
    expect(Rule.string()).toBeInstanceOf(Rule)
  })

  test('clones rule when changing generics', () => {
    const baseRule = new Rule()
    const typed = baseRule.type('string')
    expect(baseRule).not.toBe(typed)
  })

  test('clones rule when changing type-specific rules', () => {
    const baseRule = Rule.string()
    const specific = baseRule.min(5)
    expect(baseRule).not.toBe(specific)
  })

  test('returns arrays of errors/warnings by default', async () => {
    await expect(Rule.string().validate(123)).resolves.toMatchSnapshot()
  })

  test('can demote errors to warnings', async () => {
    const result = await Rule.string().warning().validate(123)

    expect(result).toMatchSnapshot()
  })

  test('can customize error messages', async () => {
    const result = await Rule.string().error('Dude it needs to be a string').validate(123)

    expect(result).toMatchSnapshot()
  })

  test('can customize warning messages', async () => {
    const result = await Rule.string().warning('Dude it should probably be a string').validate(123)

    expect(result).toMatchSnapshot()
  })

  test('can merge rules', async () => {
    const rule = new Rule().required()
    const stringRule = Rule.string().min(5)
    const newRule = rule.merge(stringRule)

    expect(rule).not.toBe(stringRule)
    expect(newRule).not.toBe(stringRule)
    expect(rule).not.toBe(newRule)

    const result = await newRule.validate('Hei')
    expect(result).toMatchSnapshot()
  })

  test('can merge rules', () => {
    expect(new Rule().required().isRequired()).toBe(true)
    expect(new Rule().isRequired()).toBe(false)
    expect(new Rule().optional().isRequired()).toBe(false)
  })
})
