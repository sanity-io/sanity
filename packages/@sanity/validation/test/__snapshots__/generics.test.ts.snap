// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generics can customize error messages 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Dude it needs to be a string",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`generics can customize info messages 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Dude it should probably be a string",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "info",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`generics can customize warning messages 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Dude it should probably be a string",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "warning",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`generics can demote errors to warnings 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Expected type \\"String\\", got \\"Number\\"",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "warning",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`generics can merge rules 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be at least 5 characters long",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`generics returns arrays of errors/warnings by default 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Expected type \\"String\\", got \\"Number\\"",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;
